server/src/main/java/org/elasticsearch/common/xcontent/support/XContentMapValues.java

```diff
@@ -297,20 +315,26 @@ private static int step(CharacterRunAutomaton automaton, String key, int state)
                 }
                 Map<String, Object> filteredValue = filter((Map<String, ?>)value,
                         includeAutomaton, includeState, excludeAutomaton, excludeState, matchAllAutomaton);
-                if (filteredValue.isEmpty() == false) {
+                if (null != filteredValue) {
                     filtered.add(filteredValue);
                 }
-            } else if (value instanceof Iterable) {
-                List<Object> filteredValue = filter((Iterable<?>) value,
+            } else if (value instanceof List) {
+                List<Object> filteredValue = filter((List<?>) value,
                         includeAutomaton, initialIncludeState, excludeAutomaton, initialExcludeState, matchAllAutomaton);
-                if (filteredValue.isEmpty() == false) {
+                if (null!=filteredValue) {
```

Comment:
this was part of the bug causing empty lists to disappear from _source. The same would have been true of map (see L300).

The filter map and filter list methods now return null to tell the parent to remove them.